doc/developer/design/20230615_webhook_source.md

```diff
@@ -0,0 +1,232 @@
+`storage-client` to create a new collection, and finally return a well formatted URL to the user.
+
+When we receive a event we'll use the created `SessionClient` to send the raw `headers` and `body`
+to the coordinator, pack the data into a row, and append it to the necessary collection using a
+new monotonic append API on the `StorageController`.
+
+## Detailed description
+
+We'll update the parser to support creating a new kind of Source, `HTTP REQUESTS`, for which you can
+specify a `FORMAT`, which denotes how we decode the body of a request, the only supported format
+for the time being is `BYTES`.
+
+```
+CREATE SOURCE <name> FROM HTTP REQUESTS FORMAT BYTES;
+```
+
+After executing this request, we'll create a source with the following columns:
+
+| name    | type              |
+|---------|-------------------|
+| headers | map[text -> text] |
+| body    | bytea             |
+
+Using `bytea` might not be optimal from a storage perspective (e.g. no chance for MFP Pushdown) but
+it makes this initial attempt at webhooks maximally compatible. On the completion of the statement
+we'll return an ID to the user, which they would specify in the webhook URL. The returned ID would
+be 6 character base62 encoded string, a pattern commonly used in URL shorteners.
+
+> **_Note_**: An alternative to `bytea` encoded bodies would be to use `jsonb`. Looking through a
+few common apps that support webhooks, the payloads all default to JSON, or have JSON as the only
+option. Using `bytea` makes us compatible with all webhooks, with the tradeoff of adding work for
+the user to cast to their desired serialization format downstream.
+
+We'll add a new endpoint to the existing [`base_router`](https://github.com/MaterializeInc/materialize/blob/6e1f4c7352427301d782438d614feafb0f644442/src/environmentd/src/http.rs#L747)
+in `environmentd` with the path: `/api/webhook/:id`. This follows the existing pattern of our other
+two endpoints, whose paths also start with `/api`.
+
+> **_Note_**: The existing Web Sockets API is at `/api/experimental/ws`, we could move the webhooks
+API under the path `/api/experimental/webhook`, if we think there will be significant breaking
+changes in the future.
+
+We'll also add a new API `SessionClient::append_webhook` which is very similar to the existing
+[`insert_rows`](https://github.com/MaterializeInc/materialize/blob/6e1f4c7352427301d782438d614feafb0f644442/src/adapter/src/client.rs#L392-L397).
+This new API will send the captured `:id` from the URL path, the headers and body of the request,
+straight to the coordinator, via a new `Command`. In the coordinator we'll do the following:
+
+1. Spawn a new `tokio::task` to do the work off the main thread.
+2. Map the `id` to a previously created Source.
+3. Remove sensitive header values from the received headers (e.g. [Authorization](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Authorization)).
```

Comment:
Ah, I left another comment somewhere without having read this line. I don't know that there's a good standard list of headers to filter out. Webhook providers don't typically set `Authorization` headers. They do usually provide a signature that you can use to verify the request. Exposing that signature isn't actually scary though, because you can only use it to validate the request; you can't use it to sign other requests...

Reply:
I think it's possible the right thing to do here is just ... no filtering.

Reply:
Great! I updated the doc and removed the part about filtering out sensitive headers. I think the approach of not including them by default, and having to opt-in is much better.